core/src/socket.rs

```diff
@@ -0,0 +1,243 @@
+use crate::{
+    avm1::{Activation, ActivationIdentifier, ExecutionReason, Object, TObject},
+    backend::navigator::NavigatorBackend,
+    context::UpdateContext,
+    string::AvmString,
+};
+use encoding_rs::UTF_8;
+use gc_arena::{Collect, CollectionContext};
+use generational_arena::{Arena, Index};
+use std::collections::VecDeque;
+
+/// XMLSocket backend implementation
+///
+/// When `XMLSocket.connect()` is called, the backend may provide a
+/// [XmlSocketConnection] instance, that will be polled every tick.
+/// Methods must not block.
+///
+/// For performances reasons, only 1 message will be polled per tick per connection
+/// and at most 1 message will be sent per tick per connection.
+///
+/// When the XMLSocket gets closed, the underlying [XmlSocketConnection]
+/// ise dropped, so simply implement [Drop] to handle cleanup stuff.
+///
+/// See [XmlSocketConnection::is_connected] for details about how
+/// Ruffle infer connection state, [XmlSocketConnection::send] for details
+/// about forwarding messages, [XmlSocketConnection::poll] for details
+/// about polling incoming messages.
+pub trait XmlSocketConnection {
+    /// Return socket connection state
+    ///
+    /// Possibles values:
+    ///  * `None`: connecting ([XmlSocketConnection::send] and
+    ///            [XmlSocketConnection::poll] not yet called)
+    ///  * `Some(true)`: connected ([XmlSocketConnection::send] and
+    ///                 [XmlSocketConnection::poll] got called)
+    ///  * `Some(false)`: disconnected, or connection refused socket wasn't connected
+    ///                   (called only after [XmlSocketConnection::poll] return `None`;
+    ///                    [XmlSocketConnection::send] and [XmlSocketConnection::poll]
+    ///                    won't be called after `Some(false)` is returned)
+    ///
+    /// Returning `None` after `Some(true)` was returned
+    /// will be considered as `Some(false)` (i.e. connection closed).
+    fn is_connected(&self) -> Option<bool>;
+
+    /// Send a message to the remote side
+    ///
+    /// `buf` contains all bytes for the message to send, it's up to the backend to
+    /// decide how to encode and transport them.
+    ///
+    /// Only called if there is at least 1 pending message to send.
+    fn send(&mut self, buf: Vec<u8>);
```

Comment:
This is probably better to be taking a `&[u8]` instead of `Vec<u8>`.

Reply:
Actually, I reread the code and I see why it takes a `Vec<u8>`, but I'm still curious whether this could be made more efficient. In the future we will be having classes like the AS3 [Socket](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/net/Socket.html), which we'll most likely want to reuse this backend in someway. In the AS3 Socket class, methods like `writeBytes` would really benefit from being able to write the bytes directly into the socket (or at least some buffer in the socket).

Reply:
First of all, thanks for the review :+1:

> This is probably better to be taking a &[u8] instead of Vec<u8>.

Are you suggesting to directly and synchronously call `XmlSocketConnection::send` when `XMLSocket.send` is called? The send buffering would have to be transferred to the backend but it's doable :+1:

> which we'll most likely want to reuse this backend in someway. 

I haven't moved the packets aggregation in the `core` part because I thought it would be better to let backends deal with that (especially for the web backend, so that it's easier to emulate servers and to allow passing buffers as string messages in `WebSocket.send` to see them in browers' devtools). To sum up, the current implementation is XMLSocket focused, so if we should support any TCP emulation mechanism, I should probably refactor it now :confused:

Also, I wasn't aware of AS3 Socket class, thanks again :+1: